SparseGrids/tsgGridWavelet.cpp

```diff
@@ -331,6 +366,37 @@ double GridWavelet::evalIntegral(const int p[]) const{
     }
     return v;
 }
+void GridWavelet::evalDiffBasis(const int p[], const double x[], double jacobian[]) const {
+    // Evaluates the derivative of the wavelet basis given at point p at the coordinates given by x.
+    // The derivative function of a symmetric function will be antisymmetric across the origin for every x != 0.0.
+    std::vector<double> value_cache(num_dimensions), derivative_cache(num_dimensions);
+    for(int i=0; i<num_dimensions; i++) {
+        value_cache[i] = rule1D.eval<0>(p[i], x[i]);
+        jacobian[i] = rule1D.eval<1>(p[i], x[i]);
+    }
+
+    // Note that excessive branching is done to mirror the logic in evalBasis() and evalIntegral().
+    int back_idx = num_dimensions-2;
+    double t = 1.0;
+    for(int i=1; i<num_dimensions; i++){
+        t *= value_cache[i-1];
+        jacobian[i] *= t;
+        if (t == 0.0) {
```

Comment:
If `t` becomes zero in the first loop, you need to zero out `jacobian` for the entries after the `back_index`.

In the case of eval, breaking on a zero value will save several calls to `evalBasis()` and each call consists of multiple if-statements and index and floating point operations. If you break early here, you will save on couple of multiplications, which is cheap already. We can (and should) add differentiation to the sparse grid benchmark to test this, but I think the early break will not help and may even hurt performance.

I haven't done any PRs lately, let me add differentiation to the benchmark so we can test this. We can talk about this later in a call.

Reply:
Read the long paragraph, in the other two cases you're saving big, here you're saving very little if anything at all.

Wait for the benchmark and we will test it.

Reply:
I see a lot of run-to-run variability when using 200 evals. Removing the early breaks does seem to make things a bit better, though.

I'll bump to 1000 repetitions.

Reply:
Same result, I see a slight bump in performance if I remove the early breaking.

Reply:
Gotcha. I've removed the early breaking in the latest push.

Reply:
Okay. Let me run it a few times.

Reply:
I've tested it a few times on my home PC with an Intel(R) Core(TM) i7-10700K CPU, and there is only the slightest margin of improvement if I remove early breaking.

Here is a sample output with breaking:
```
./benchmarksgrid differentiate wavelet 4 1 2 level rleja 3 2000 5
differentiate wavelet 4 1 2 level rleja 3 2000 5 (points: 9025)
    note: reporting total time (ms), does not normalize (divide) by iterations
   3374
   3416
   3448
   3758
   4262
```
and here is one without breaking:
```
./benchmarksgrid differentiate wavelet 4 1 2 level rleja 3 2000 5
differentiate wavelet 4 1 2 level rleja 3 2000 5 (points: 9025)
    note: reporting total time (ms), does not normalize (divide) by iterations
   3246
   3327
   3473
   3745
   4225
```
Based on the results, I'll leave out the early breaking.

Reply:
Tested on Ryzen, no change detected. If there's no difference, then we keep the simpler code.

The CI system has an issue, I will manually test tomorrow and merge.